File: sources/filesystem/generate_files.py
"""
Filesystem Source - Generates daily CSV and Parquet files for Mar 3-9, 2026.
Simulates an additional sales channel (e-commerce / EDI exports).
"""

import os
import random
from datetime import date, timedelta
from pathlib import Path

import pandas as pd

CITIES = [
    ("São Paulo", "SP"), ("Rio de Janeiro", "RJ"), ("Belo Horizonte", "MG"),
    ("Curitiba", "PR"), ("Porto Alegre", "RS"), ("Salvador", "BA"),
    ("Fortaleza", "CE"), ("Recife", "PE"), ("Manaus", "AM"), ("Brasília", "DF"),
    ("Goiânia", "GO"), ("Campo Grande", "MS"), ("Florianópolis", "SC"),
    ("Belém", "PA"), ("São Luís", "MA"), ("Maceió", "AL"), ("Vitória", "ES"),
    ("Campinas", "SP"), ("Santos", "SP"),
]

PRODUCTS = [
    (101, "Notebook Dell Inspiron 15",      "Eletrônicos",      3499.90),
    (102, "Smartphone Samsung Galaxy A54",  "Eletrônicos",      1799.90),
    (103, "Monitor LG 27\" 4K",             "Eletrônicos",      2199.90),
    (104, "Smart TV Samsung 55\" QLED",     "Eletrônicos",      4299.90),
    (105, "Tablet iPad Air 5",              "Eletrônicos",      4999.90),
    (106, "Teclado Mecânico Redragon",       "Periféricos",       399.90),
    (107, "Mouse Logitech MX Master 3",     "Periféricos",       599.90),
    (108, "Headset JBL Quantum 400",        "Áudio",             449.90),
    (109, "Cadeira Gamer ThunderX3",        "Móveis",           1299.90),
    (110, "Ar Condicionado Daikin 12k",     "Climatização",     3199.90),
]

SALESMEN = list(range(1, 21))

random.seed(42)

OUTPUT_DIR = Path(__file__).parent / "data"
OUTPUT_DIR.mkdir(exist_ok=True)

START_DATE = date(2026, 3, 3)
END_DATE   = date(2026, 3, 9)


def generate_day(day: date) -> pd.DataFrame:
    is_weekend = day.weekday() >= 5
    n_records = random.randint(8, 12) if is_weekend else random.randint(20, 35)

    rows = []
    for _ in range(n_records):
        product = random.choice(PRODUCTS)
        city, state = random.choice(CITIES)
        salesman_id = random.choice(SALESMEN)
        quantity = random.randint(1, 6)
        rows.append({
            "source":       "filesystem",
            "sale_date":    day.isoformat(),
            "salesman_id":  salesman_id,
            "product_id":   product[0],
            "product_name": product[1],
            "category":     product[2],
            "city":         city,
            "state":        state,
            "quantity":     quantity,
            "unit_price":   product[3],
            "total_amount": round(quantity * product[3], 2),
        })
    return pd.DataFrame(rows)


def main():
    current = START_DATE
    while current <= END_DATE:
        df = generate_day(current)
        date_str = current.strftime("%Y%m%d")

        # CSV export
        csv_path = OUTPUT_DIR / f"sales_{date_str}.csv"
        df.to_csv(csv_path, index=False)
        print(f"[OK] CSV  → {csv_path}  ({len(df)} records)")

        # Parquet export (partitioned)
        parquet_path = OUTPUT_DIR / f"sales_{date_str}.parquet"
        df.to_parquet(parquet_path, index=False)
        print(f"[OK] Parquet → {parquet_path}  ({len(df)} records)")

        current += timedelta(days=1)

    print("\nAll files generated successfully.")


if __name__ == "__main__":
    main()

File: src/main/scala/com/datakata/datagen/GenerateFiles.scala
package com.datakata.datagen

import com.datakata.config.AppConfig
import org.slf4j.LoggerFactory

import java.io.{File, PrintWriter}
import java.time.{DayOfWeek, LocalDate}

/**
 * Generates daily CSV sales files for the filesystem source.
 * Plain Scala I/O — no Spark or external dependencies required.
 *
 * Output: <FS_DATA_DIR>/sales_YYYY-MM-DD.csv  (one flat file per day)
 */
object GenerateFiles {

  private val log    = LoggerFactory.getLogger(getClass)
  private val config = AppConfig()

  private val cities = Seq(
    ("São Paulo", "SP", "Sudeste"), ("Rio de Janeiro", "RJ", "Sudeste"),
    ("Belo Horizonte", "MG", "Sudeste"), ("Curitiba", "PR", "Sul"),
    ("Porto Alegre", "RS", "Sul"), ("Salvador", "BA", "Nordeste"),
    ("Fortaleza", "CE", "Nordeste"), ("Recife", "PE", "Nordeste"),
    ("Manaus", "AM", "Norte"), ("Brasília", "DF", "Centro-Oeste"),
    ("Goiânia", "GO", "Centro-Oeste"), ("Belém", "PA", "Norte"),
    ("Florianópolis", "SC", "Sul"), ("São Luís", "MA", "Nordeste"),
    ("Maceió", "AL", "Nordeste"), ("Natal", "RN", "Nordeste"),
    ("Teresina", "PI", "Nordeste"), ("Campo Grande", "MS", "Centro-Oeste"),
    ("João Pessoa", "PB", "Nordeste")
  )

  private val products = Seq(
    (1, "Notebook Dell Inspiron 15",     "Eletrônicos",   3499.90),
    (2, "Smartphone Samsung Galaxy A54", "Eletrônicos",   1799.90),
    (3, "Monitor LG 27 4K",              "Eletrônicos",   2199.90),
    (4, "Teclado Mecânico Redragon",      "Periféricos",    399.90),
    (5, "Mouse Logitech MX Master 3",    "Periféricos",    599.90),
    (6, "Cadeira Gamer ThunderX3",       "Móveis",        1299.90),
    (7, "Headset JBL Quantum 400",       "Áudio",          449.90),
    (8, "Tablet iPad Air 5",             "Eletrônicos",   4999.90),
    (9, "Impressora HP LaserJet",        "Periféricos",    899.90),
    (10,"SSD Kingston 1TB",              "Armazenamento",  399.90)
  )

  private val salesmen = (1 to 20).map(i => (i, s"Vendedor $i"))
  private val rng      = new scala.util.Random(42)

  private val HEADER =
    "source,sale_date,salesman_id,salesman_name,product_id,product_name," +
    "category,city,state,region,quantity,unit_price,total_amount"

  def main(args: Array[String]): Unit = {
    val outDir = new File(config.fsDataDir)
    outDir.mkdirs()

    val start = LocalDate.parse(config.dataStartDate)
    val end   = LocalDate.parse(config.dataEndDate)

    var date = start
    while (!date.isAfter(end)) {
      val file = new File(outDir, s"sales_$date.csv")
      val pw   = new PrintWriter(file, "UTF-8")
      try {
        pw.println(HEADER)
        generateDay(date).foreach(pw.println)
      } finally {
        pw.close()
      }
      log.info(s"Generated $file")
      date = date.plusDays(1)
    }

    log.info(s"Data generation complete → ${outDir.getAbsolutePath}")
  }

  private def generateDay(date: LocalDate): Seq[String] = {
    val isWeekend = date.getDayOfWeek == DayOfWeek.SATURDAY ||
                    date.getDayOfWeek == DayOfWeek.SUNDAY
    val count = if (isWeekend) 8 + rng.nextInt(5) else 20 + rng.nextInt(16)

    (1 to count).map { _ =>
      val (pid, pname, cat, price) = products(rng.nextInt(products.size))
      val (sid, sname)             = salesmen(rng.nextInt(salesmen.size))
      val (city, state, region)    = cities(rng.nextInt(cities.size))
      val qty                      = 1 + rng.nextInt(3)
      val total                    = price * qty
      s"filesystem,$date,$sid,$sname,$pid,$pname,$cat,$city,$state,$region,$qty,$price,$total"
    }
  }
}
